doc: notation supports OCI Image manifest (#217)
notaryproject/notaryproject: specs/signature-specification.md

## Storage

This section describes how Notary v2 signatures are stored in the OCI Distribution conformant registry.

> Notary v2 uses [OCI artifact manifest][oci-artifact-manifest] to store the signature in the registry.

Reviewer: ```suggestion
Notary v2 uses [OCI artifact manifest][oci-artifact-manifest] to store signatures in the registry.
```

  - **`io.cncf.notary.x509chain.thumbprint#S256`**: A REQUIRED annotation whose value contains the list of SHA-256 fingerprint of signing certificate and certificate chain (including root) used for signature generation. The annotation name contains the hash algorithm as a suffix (`#S256`) and can be extended to support other hashing algorithms in future.
    The list of fingerprints is present as a JSON array string, corresponding to ordered certificates in [*Certificate Chain* unsigned attribute](#unsigned-attributes) in the signature envelope.

### Backwards Compatibility

> Notary v2 MAY support using [OCI Image manifest][oci-image-manifest] to store the signature in the registries that implement partial or older versions of the OCI Image specification.

Reviewer 2: This statement is not quite accurate. The `subject` field for the OCI image manifest is introduced in `image-spec v1.1`. Therefore, registries supporting OCI image-spec v1.0 may still not support OCI image manifest with the `subject` field.

Reviewer 3: > This statement is not quite accurate. The `subject` field for the OCI image manifest is introduced in `image-spec v1.1`. Therefore, registries supporting OCI image-spec v1.0 may still not support OCI image manifest with the `subject` field.

when pushing the fallback manifest, oras library handles the complexities of creating the relationship between artifact and subject via fallback mechanism.

Reviewer 4: @shizhMSFT - I think if registries don't support additional properties like the subject, I think its fair to say that the it doesn't support extended fields for even Image-spec 1.0 and maybe we should not handle that. For e.g. example the registry might not even support index. From the tool perspective do you think having the subject property in v1.0 is ok?

Reviewer 2: From the tool perspective, having the `subject` property in `v1.0` is definitely ok since it is indeed partial `v1.1`.

My focus is the accuracy of the wording in the spec documentation. That is, implementing `image-spec v1.0` does not imply supporting the `subject` field.

How about
```suggestion
Notary v2 MAY support using [OCI image manifest][oci-image-manifest] to store the signature in the registries that implement partial OCI Image specification v1.1.
```

Author: @shizhMSFT How about I update the link of `oci-image-manifest` to `OCI image manifest v1.1.0-rc.2` to indicate that notation using OCI image manifest with `subject` field, while keep the wording `implement partial or older versions of the OCI Image specification. ` Because with your suggestion, it seems registries implementing v1.0 are excluded. I think some registries implementing v1.0 support `subject` field, since it is an optional field.

Reviewer 2: > some registries implementing v1.0 support `subject` field

Yes, you may consider those registries partially v1.1 compatible.

> Signature Manifest example per OCI Image manifest:

Reviewer 2: nit:
```suggestion
Signature manifest example per OCI image manifest:
```

```jsonc
{
    "schemaVersion": 2,
    "mediaType": "application/vnd.oci.image.manifest.v1+json",
    "config": {

>         "mediaType": "application/vnd.oci.image.config.v1+json",

Reviewer 2: The media type of the config layer is used as the artifact type. Therefore,
```suggestion
        "mediaType": "application/vnd.cncf.notary.signature",
```

        "size": 2,
        "digest": "sha256:ca3d163bab055381827226140568f3bef7eaac187cebd76878e0b63e9e442356"
    },
    "layers": [
        {
            "mediaType": "application/jose+json",
            "digest": "sha256:9834876dcfb05cb167a5c24953eba58c4ac89b1adf57f28f2f9d09af107ee8f0",
            "size": 32654
        }
    ],
    "subject": {
        "mediaType": "application/vnd.oci.image.manifest.v1+json",
        "digest": "sha256:73c803930ea3ba1e54bc25c2bdc53edd0284c62ed651fe7b00369da519a3c333",
        "size": 16724
    },
    "annotations": {
        "io.cncf.notary.x509chain.thumbprint#S256": 
        "[\"B7A69A70992AE4F9FF103EBE04A2C3BA6C777E439253CE36562E6E98375068C3\",\"932EB6F5598435D4EF23F97B0B5ACB515FAE2B8D8FAC046AB813DDC419DD5E89\"]"
    }
}
```

Besides the [image manifest property requirements][image-manifest-property-descriptions], the properties has the following additional restrictions:

- **`mediaType`** (*string*): This REQUIRED property MUST be `application/vnd.oci.image.manifest.v1+json`.

> - **`config`** (*descriptor*): This REQUIRED property references a configuration object for a signature by digest.

Reviewer 2: The content of the `config` is not specified.

  - **`io.cncf.notary.x509chain.thumbprint#S256`**: A REQUIRED annotation whose value contains the list of SHA-256 fingerprint of signing certificate and certificate chain (including root) used for signature generation. The annotation name contains the hash algorithm as a suffix (`#S256`) and can be extended to support other hashing algorithms in future.
    The list of fingerprints is present as a JSON array string, corresponding to ordered certificates in [*Certificate Chain* unsigned attribute](#unsigned-attributes) in the signature envelope.

### Backward Compatibility

> Notary v2 MAY support using `OCI image manifest` to store the signature in the registries that implement partial of the [OCI image specification v1.1][oci-image-manifest].

Reviewer: ```suggestion
Notary v2 MAY support using `OCI image manifest` to store signatures in registries that implement partial of the [OCI image specification v1.1][oci-image-manifest].
```

Signature manifest example per OCI image manifest:

```jsonc
{
    "schemaVersion": 2,
    "mediaType": "application/vnd.oci.image.manifest.v1+json",
    "config": {
        "mediaType": "application/vnd.cncf.notary.signature",
        "size": 2,
        "digest": "sha256:44136fa355b3678a1146ad16f7e8649e94fb4fc21fe77e8310c060f61caaff8a"
    },
    "layers": [
        {
            "mediaType": "application/jose+json",
            "digest": "sha256:9834876dcfb05cb167a5c24953eba58c4ac89b1adf57f28f2f9d09af107ee8f0",
            "size": 32654
        }
    ],
    "subject": {
        "mediaType": "application/vnd.oci.image.manifest.v1+json",
        "digest": "sha256:73c803930ea3ba1e54bc25c2bdc53edd0284c62ed651fe7b00369da519a3c333",
        "size": 16724
    },
    "annotations": {
        "io.cncf.notary.x509chain.thumbprint#S256": 
        "[\"B7A69A70992AE4F9FF103EBE04A2C3BA6C777E439253CE36562E6E98375068C3\",\"932EB6F5598435D4EF23F97B0B5ACB515FAE2B8D8FAC046AB813DDC419DD5E89\"]"
    }
}
```

> Besides the [image manifest property requirements][image-manifest-property-descriptions], the properties has the following additional restrictions:

Reviewer: ```suggestion
Besides the [image manifest property requirements][image-manifest-property-descriptions], the property has the following additional restrictions:
```

Author: I would like to keep the plural and changed the word `has` to `have`.

- **`mediaType`** (*string*): This REQUIRED property MUST be `application/vnd.oci.image.manifest.v1+json`.

> - **`config`** (*descriptor*): This property is REQUIRED to be compatible with [OCI image specification][oci-image-manifest]. Notary v2 doesn't require any configurations for a signature, and the configuration contents are not consumed by Notary v2.

Reviewer: ```suggestion
- **`config`** (*descriptor*): This property is REQUIRED to be compatible with [OCI image specification][oci-image-manifest]. Notary v2 doesn't require any configuration for a signature, and the configuration content is not consumed by Notary v2.
```

> - **`config`** (*descriptor*): This REQUIRED property references a configuration object for a signature by digest. Notary v2 doesn't require specific configurations for a signature. Implementations SHOULD decide the configuration content based on their specific needs.

Reviewer 2: Should we call out that the content of `config` is never used by notary v2 as it exists just because of compatibility?

Reviewer 2: Note: There is no config in the original OCI Artifact Manifest.

Reviewer 2: If we store meaningful config in the OCI image manifest, then we are not able to convert OCI image manifest to OCI artifact manifest and vice versa in the future.

>   - **`mediaType`** (*string*): This REQUIRED property MUST be `application/vnd.cncf.notary.signature`.

Reviewer 5: If the config is not consumed by notary then are we enforcing mediaType to be  `application/vnd.cncf.notary.signature` just for sanity check? Should notary implementation check the value?

Author: This is defined by the spec here 
https://github.com/opencontainers/distribution-spec/blob/main/spec.md#pushing-manifests-with-subject

and referrer API also requires https://github.com/opencontainers/distribution-spec/blob/main/spec.md#listing-referrers

  - **`digest`** (*string*): This REQUIRED property is the digest of the config content.

>   - **`size`** (*int64*): This REQUIRED property specifies the size, in bytes, of the raw config content.

Reviewer 5: What should be values of these two fields?

Author: This is the spec. The values leave to the decision of different implementations.

> - **`layers`** (*array of objects*): This REQUIRED property contains collection of only one [OCI descriptor][oci-descriptor] referencing signature envelope.

Reviewer: ```suggestion
- **`layers`** (*array of objects*): This REQUIRED property contains collection of only one [OCI descriptor][oci-descriptor] referencing the signature envelope.
```

[ietf-rfc3161]: https://datatracker.ietf.org/doc/html/rfc3161#section-2.4.2
[oci-artifact-manifest]: https://github.com/opencontainers/image-spec/blob/main/artifact.md
[oci-distribution-referrers]: https://github.com/opencontainers/distribution-spec/blob/main/spec.md#listing-referrers
[oci-image-manifest]: https://github.com/opencontainers/image-spec/blob/main/manifest.md

> [image-manifest-property-descriptions]: https://github.com/opencontainers/image-spec/blob/main/manifest.md#image-manifest-property-descriptions

Reviewer 2: Specific version of the spec should be referenced instead of `main`.

> [annotation-rules]: https://github.com/opencontainers/image-spec/blob/main/annotations.md#rules
> [oci-descriptor]: https://github.com/opencontainers/image-spec/blob/main/descriptor.md
> [ietf-rfc3161]: https://datatracker.ietf.org/doc/html/rfc3161#section-2.4.2

Reviewer: Somehow, I cannot see these links in the webpage. I think they are not successfully rendered in the markdown? hence:
```suggestion
## References
annotation-rules: https://github.com/opencontainers/image-spec/blob/main/annotations.md#rules</br>
oci-descriptor: https://github.com/opencontainers/image-spec/blob/main/descriptor.md</br>
ietf-rfc3161: https://datatracker.ietf.org/doc/html/rfc3161#section-2.4.2</br>
```

Author: This is the correct usage for replacing the reference with the links for a markdown file. So I will keep it unchanged.

> [oci-artifact-manifest]: https://github.com/opencontainers/image-spec/blob/v1.1.0-rc2/artifact.md
> [oci-distribution-referrers]: https://github.com/opencontainers/distribution-spec/blob/v1.1.0-rc1/spec.md#listing-referrers
> [oci-image-manifest]: https://github.com/opencontainers/image-spec/blob/v1.1.0-rc2/manifest.md
> [image-manifest-property-descriptions]: https://github.com/opencontainers/image-spec/blob/v1.1.0-rc2/manifest.md#image-manifest-property-descriptions

Reviewer: Same here:
```suggestion
oci-artifact-manifest: https://github.com/opencontainers/image-spec/blob/v1.1.0-rc2/artifact.md</br>
oci-distribution-referrers: https://github.com/opencontainers/distribution-spec/blob/v1.1.0-rc1/spec.md#listing-referrers</br>
oci-image-manifest: https://github.com/opencontainers/image-spec/blob/v1.1.0-rc2/manifest.md</br>
image-manifest-property-descriptions: https://github.com/opencontainers/image-spec/blob/v1.1.0-rc2/manifest.md#image-manifest-property-descriptions</br>
```

Author: This is the correct usage for replacing the reference with the links for a markdown file. So I will keep it  unchanged.
